canopen_ros2_control/include/canopen_ros2_control/canopen_system.hpp

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) 2022, StoglRobotics
+// Copyright (c) 2022, Stogl Robotics Consulting UG (haftungsbeschränkt) (template)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+//----------------------------------------------------------------------
+/*!\file
+ *
+ * \author  Lovro Ivanov lovro.ivanov@gmail.com
+ * \date    2022-06-29
+ *
+ */
+//----------------------------------------------------------------------
+
+#ifndef CANOPEN_ROS2_CONTROL__CANOPEN_SYSTEM_HPP_
+#define CANOPEN_ROS2_CONTROL__CANOPEN_SYSTEM_HPP_
+
+#include <string>
+#include <vector>
+
+#include "canopen_ros2_control/visibility_control.h"
+#include "hardware_interface/system_interface.hpp"
+#include "hardware_interface/handle.hpp"
+#include "hardware_interface/hardware_info.hpp"
+#include "hardware_interface/types/hardware_interface_return_values.hpp"
+#include "rclcpp/macros.hpp"
+#include "rclcpp_lifecycle/state.hpp"
+
+#include <rclcpp/executors.hpp> // for MultiThreadedExecutor
+#include <canopen_core/device_manager.hpp> // for DeviceManager
+
+
+
+namespace canopen_ros2_control
+{
+    // needed auxiliary struct for ros2 control double registration
+    struct Ros2ControlCOData{
+        ros2_canopen::COData original_data;
+
+        double index; // cast to uint16_t
+        double subindex; // cast to uint8_t
+        double data;    // cast to uint32_t
+        double type;// cast to uint8_t
+
+    };
+
+    struct RORos2ControlCOData: public Ros2ControlCOData{
+
+        void set_data(ros2_canopen::COData d){
+            original_data = d;
+
+            index = static_cast<double>(original_data.index_);
+            subindex = static_cast<double>(original_data.subindex_);
+            data = static_cast<double>(original_data.data_);
+            type = static_cast<double>(original_data.type_);
+
+        }
+    };
+
+    struct WORos2ControlCoData: public Ros2ControlCOData{
+
+        WORos2ControlCoData(): one_shot(std::numeric_limits<double>::infinity()){}
+
+        // needed internally for write-only data
+        double one_shot;
+
+        bool write_command(){
+            bool ret_val;
+            // store ret value
+            ret_val = (one_shot != std::numeric_limits<double>::infinity());
+            // reset the existing active command if one exists
+            one_shot = std::numeric_limits<double>::infinity();
```

Comment:
Would this also work?

```suggestion
            one_shot = std::numeric_limits<double>::quietNaN();
```